Tour mode -- again

zneedell opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/LBNL-UCB-STI/beam/3900)
<!-- Reviewable:end -->

zneedell committed:
tour mode squashed

zneedell committed:
fix tests

zneedell committed:
fmt

zneedell committed:
start keeping track of tour_id in beamPlan (NOT WORKING!)

zneedell committed:
start keeping track of tour_id in beamPlan (fixed)

zneedell committed:
First successful test

zneedell committed:
second test

zneedell committed:
third test (failing)

zneedell committed:
fixed test for when car_based tour is input

zneedell committed:
fixed test for when no tour mode exists

zneedell committed:
ignore remaining messages after test is passed

zneedell committed:
Add test for using shared car on walk_based tour

zneedell committed:
partial fix for personalVehiclePlanSpec

zneedell committed:
allow nested subtours

zneedell committed:
minor refactoring in multinomial logit

zneedell committed:
Allow agent to use vehicle from parent tour in subtour

zneedell committed:
Fixes to allow agent to use vehicle from parent tour after completion of subtour

zneedell committed:
fixes to PersonWithPersonalVehiclePlanSpec

zneedell committed:
don't ask shared fleet managers for vehicles if we're already on a vehicle based tour

zneedell committed:
don't choose shared vehicles as tour vehicles

dimaopen committed:
Merge branch 'refs/heads/develop' into tour-mode-clean

# Conflicts:
#	production/austin
#	production/sfbay
#	src/main/scala/beam/agentsim/agents/PersonAgent.scala
#	src/main/scala/beam/agentsim/agents/household/HouseholdFleetManager.scala
#	src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala
#	src/main/scala/beam/agentsim/infrastructure/taz/TAZTreeMap.scala
#	src/main/scala/beam/router/skim/core/RideHailSkimmer.scala

dimaopen committed:
Fixes after merging develop to tour-mode-clean

zneedell committed:
replanning fixes

zneedell committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into tour-mode-clean-develop202403

# Conflicts:
#	production/seattle
#	production/sfbay
#	src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala
#	src/main/scala/beam/router/skim/readonly/ODSkims.scala

zneedell committed:
Merge remote-tracking branch 'origin/develop' into tour-mode-clean-develop202403

zneedell committed:
fix test

zneedell committed:
fix test

zneedell committed:
bugfix for including RH vehicles in main routingRequest

zneedell committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into tour-mode-clean-develop202403

zneedell committed:
tricky fix for ActivitiesDurationSpec

zneedell committed:
fix some tests

zneedell committed:
Better handling for vehicles and tour modes during replanning

zneedell commented:
Test now!

zneedell committed:
fixes for tour vehicles with drive_transit

zneedell committed:
Fixes to emergency stalll generation in test

zneedell committed:
Fixes to emergency stalll generation in test

zneedell committed:
fix to account for walk access/egress to vehicles

zneedell commented:
test now!

zneedell committed:
more test fixes

zneedell committed:
clean up

zneedell committed:
fix some mroe tests

zneedell commented:
test now!

zneedell committed:
fix some mroe tests

zneedell committed:
fix network relaxation test by making sure (emergency) vehicles appear at household

zneedell commented:
test now!

zneedell committed:
fix plan writing in tests

zneedell committed:
add freight tour mode

zneedell commented:
test now!

zneedell committed:
Fixes for freight and for plans that don't make sense

zneedell commented:
Test!

zneedell committed:
Fix test to account for additional replanning event

zneedell committed:
Check for too-short parking durations before deciding to charge

zneedell committed:
Fix for sending double LeavingParkingEvents with short trips

zneedell commented:
Test!

zneedell committed:
test fix for spurious leaving parking event

zneedell committed:
Fix for parking durations

zneedell commented:
Test!

zneedell committed:
Fix for returning vehicles when not used

zneedell committed:
update ModeChoiceSpec

zneedell commented:
Test!

zneedell committed:
Small fixes

zneedell commented:
Test!

zneedell committed:
fix for keeping vehicle after replanning during drive-transit trip

zneedell committed:
test and shared vehicle fixes

zneedell commented:
Test now!

zneedell committed:
Start refactoring by extracting some methods. Goal is to move tour mode choice after initial routing requests so they can be based on actual vehicle availability

zneedell commented:
Test!

zneedell committed:
Move tour mode choice after initial routing requests

zneedell commented:
Test now!

zneedell committed:
partial test fixes

zneedell committed:
fix for retaining vehicle in nested tour

zneedell commented:
Test now!

zneedell committed:
debug messages

zneedell commented:
Test now!

zneedell committed:
Fix for missing teleportation vehicles

zneedell committed:
minor cleanup

zneedell commented:
Test!

zneedell commented:
Test now!

dimaopen commented on src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala:
```diff
@@ -272,298 +251,123 @@ trait ChoosesMode {
+      val currentTourStrategy = _experiencedBeamPlan.getTourStrategy[TourModeChoiceStrategy](nextAct)
+      val currentTripStrategy = _experiencedBeamPlan.getTripStrategy[TripModeChoiceStrategy](nextAct)
+      if (currentTripStrategy.mode != personData.currentTripMode) {
+        logger.debug("Why are the trip mode from the beamPlan and personData different?")
```
Maybe we need to add more debug info here? personId, activity index, strategies we got, currentTripMode etc

dimaopen commented on src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala:
```diff
@@ -124,115 +127,91 @@ trait ChoosesMode {
+        logger.error(
+          s"Actor has trip mode $tripModeOption and tour " +
+          s"mode $tourModeOption, which shouldn't ever happen"
+        )
```
We need add personId, tick in order to be able to debug later.

dimaopen commented on src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala:
```diff
@@ -272,298 +251,123 @@ trait ChoosesMode {
+          Some(WALK_TRANSIT)
+        case _ =>
+          log.error(
+            "Unexpected behavior"
```
Needs person id.

dimaopen commented on src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala:
```diff
@@ -1331,90 +1295,147 @@ trait ChoosesMode {
+                  beamScenario.beamConfig.beam.agentsim.agents.vehicles.generateEmergencyHouseholdVehicleWhenPlansRequireIt =>
+              logger.warn(
+                "Ended up stuck without a car despite having car in plans, so sending the request " +
+                "back through in order to create an emergency vehicle"
```
Person id and tick or activity index

dimaopen commented on src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala:
```diff
@@ -1547,138 +1568,246 @@ trait ChoosesMode {
-      val origin = beamServices.geo.utm2Wgs(
+  when(FinishingModeChoice, stateTimeout = Duration.Zero) {
+    case Event(_: RideHailResponse, data: ChoosesModeData) =>
+      logger.warn("Recieved a ride hail response even though we'd already moved on after choosing our mode")
```
we could print out the ride-hail response for future debug. Usually unexpected messages indicates issues in the code.

dimaopen commented on src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala:
```diff
@@ -1758,7 +2359,7 @@ object ChoosesMode {
     isWithinTripReplanning: Boolean = false,
     cavTripLegs: Option[CavTripLegsResponse] = None,
-    excludeModes: Vector[BeamMode] = Vector(),
+    excludeModes: Set[BeamMode] = Set(),
```
It's better to use Set.empty in order to decrease number of objects being created.

dimaopen commented on src/main/scala/beam/agentsim/agents/PersonAgent.scala:
```diff
@@ -1243,9 +1274,14 @@ class PersonAgent(
-      log.debug("Missed transit pickup, late by {} sec", _currentTick.get - data.nextLeg.beamLeg.startTime)
+      log.debug(
+        "Agent {} missed transit pickup on {} trip, late by {} sec",
+        id.toString,
```
Don't need to call toString explicitly.

dimaopen commented on src/main/scala/beam/agentsim/agents/PersonAgent.scala:
```diff
@@ -1339,6 +1375,7 @@ class PersonAgent(
         case Some(activity) =>
           val (tick, triggerId) = releaseTickAndTriggerId()
           val activityEndTime = calculateActivityEndTime(activity, tick)
+          activity.setStartTime(tick.toDouble)
```
Doesn't it change the person plans? Why don't we also change the activity end time?

zneedell replied:
Yeah good point. This was a little bit tricky because I think it only comes up during the ActivitiesDurationSpec, where it's used by beam.agentsim.agents.PersonAgent#calculateActivityEndTime, but I'll also set end time here for consistency and confirm that it doesn't lead to any problems

dimaopen reviewed (commented):
Several minor notes.

dimaopen commented on src/main/resources/beam-template.conf:
```diff
@@ -1029,9 +1029,9 @@ beam.routing {
+      walk = "int | 1"
       car = "int | 300"
-      ride_hail = "int | 0"
+      ride_hail = "int | 30"
```
We need to mirror these changes in `inputs.rst`.

dimaopen reviewed (commented):
It's better to add some documentation about the whole feature in `behaviors.rst`.

zneedell committed:
Refactor code for consistency and encapsulation

Updated method visibility to private where appropriate. Improved variable naming and type annotations. Removed unused imports.

zneedell committed:
Address Dima's suggestions, make some methods private

zneedell committed:
Update mode choice documentation and default ride_hail cost

zneedell committed:
Refactor orphaned plug-out event handling and logging

zneedell commented:
Test now!

zneedell committed:
Fmt

zneedell commented:
Test now!

zneedell committed:
Fix for freight -- retain freight vehicle as currentTourVehicle

zneedell commented:
Test now!

dimaopen commented:
Test!

zneedell committed:
fix for building docker image on ARM64

zneedell committed:
fix for building docker image on ARM64

zneedell committed:
keep track of tour_id from urbansim plans

zneedell committed:
Merge remote-tracking branch 'origin/develop' into tour-mode-clean-develop202403

# Conflicts:
#	src/main/scala/beam/agentsim/agents/PersonAgent.scala
#	src/main/scala/beam/agentsim/agents/modalbehaviors/DrivesVehicle.scala
#	src/main/scala/beam/agentsim/events/PathTraversalEvent.scala

zneedell committed:
merge fix

zneedell commented:
Test!

zneedell committed:
Merge remote-tracking branch 'origin/develop' into tour-mode-clean-develop202403

# Conflicts:
#	src/main/scala/beam/agentsim/agents/parking/ChoosesParking.scala
#	src/test/scala/beam/agentsim/infrastructure/ZonalParkingManagerSpec.scala

zneedell committed:
merge fix

beambot2 commented:
test now!

beambot2 commented:
test now!

beambot2 commented:
test!

zneedell committed:
stuckness fixes in PILATES

zneedell committed:
fix for NoSuchElementException in AgentSimToPhysSimPlanConverter

zneedell committed:
Fixes for strange and nested plans from ASim

zneedell committed:
Merge fixes

zneedell commented:
Test now!

zneedell committed:
Fix for missing tour id

zneedell commented:
Test now!

zneedell committed:
fixes for keeping the correct tour vehicle

zneedell committed:
fix tour mode spec

zneedell committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into tour-mode-clean-develop202403

# Conflicts:
#	build.gradle

zneedell commented:
Test now!

zneedell committed:
fix for vehicle sharing spec

zneedell commented:
Test now!

zneedell committed:
test fixes

zneedell commented:
Test now!

zneedell committed:
fix flaky test

zneedell commented:
Test now!

zneedell commented:
Test now!

zneedell commented:
Test!

zneedell committed:
fix flaky test

zneedell committed:
update tutorial

zneedell committed:
Let agents see parent tour vehicle if they're going on a car_based subtour

zneedell commented:
Test now!

beambot2 commented:
test!

beambot2 commented:
test!

beambot2 commented:
test!

beambot2 commented:
test now!
